app/src/main/java/org/mozilla/fenix/ext/BrowserIcons.kt

```diff
@@ -4,8 +4,35 @@
 
 package org.mozilla.fenix.ext
 
+import android.graphics.drawable.Drawable
+import android.view.View
 import android.widget.ImageView
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
 import mozilla.components.browser.icons.BrowserIcons
 import mozilla.components.browser.icons.IconRequest
 
-fun BrowserIcons.loadIntoView(view: ImageView, url: String) = loadIntoView(view, IconRequest(url))
+fun BrowserIcons.loadIntoView(
+    view: ImageView,
+    url: String = "",
+    localImage: Drawable? = null,
+    iconCover: ImageView? = null
+): Job? {
+    if (localImage != null) view.setImageDrawable(localImage)
+    GlobalScope.launch {
+        if (url.isNotBlank()) {
+            loadIntoView(view = view, request = IconRequest(url))
+        }
+        delay(DELAY_UNTIL_REVEAL)
```

Comment:
It looks to me that we're using this delay to synchronize the setting of image with all the top sites. I don't think an arbitrary delay value would work here because devices and networks vary. Maybe we need to update the top sites adapter to show the top sites icon when all the `loadIntoView` calls return? I recall Petru doing [something similar](https://github.com/mozilla-mobile/android-components/pull/8894) for the tabs tray adapter as well that we could try. What do you think?

Reply:
Thank you for looking into this @jonalmeida.
> We use this `loadIntoView` in [a few places](https://searchfox.org/mozilla-mobile/search?q=loadIntoView&path=fenix&case=false&regexp=false) other than top sites, so this delay would affect all those cases. Is that what we want?

You are right. Sorry about that :(

 
> I don't think an arbitrary delay value would work here because devices and networks vary. Maybe we need to update the top sites adapter to show the top sites icon when all the loadIntoView calls return?

I've spent quite a bit of time in trying to implement just this, but in the end there were a lot of changes (17 files modified) and it wasn't working great. Plus it was either wait for the first topSite to load and unveil all of them OR wait for all topSites to load and unveil them then, the latter choice could break the whole system if one of the topSite icons just wouldn't load, while the former wouldn't be much of an improvement.

I put this PR together rather quickly and I apologise for its low quality. I have to admit I'm not sure if there is any satisfactory solution to this issue, at least not with my current experience level, so I'm unsure how to proceed. I'll take a look at Petru's PR, but I've already talked to him about this issue.

Reply:
I wouldn't want to hold you up on a this if we can't find a better solution, so maybe let's try to localize the solution to just the top sites then?

Not sure if you've tried this already, but instead of using `loadIntoView` how about trying `loadIcon` so you can call `awaitAll` on the icons and notify the adapter when that is complete. I wouldn't spend too much time to get this to work if it ends up being as messy as the other solution since you've already spent time on this. 🙂